### DISCUSSION TITLE
Parallel workflows

### DISCUSSION THREAD

**Discussion by justin808** [sidekiq/sidekiq#5110] [General]
Is it possible to create a top-level batch that has this sort of parallel workflow?

Currently, the overall batch sequentially runs 3 child batches. Each child batch uses the [bulk queuing](https://github.com/mperham/sidekiq/wiki/Bulk-Queueing) feature. The on_success callback triggers the next batch, so that on success for the "A jobs" batch triggers the creation of the "B jobs" batch.

I'd like to see if I could use extra batches to for below structure of creating all the A jobs, B jobs, and C jobs early.

1. We can see the batches get processed in the sidekiq batch UX.
2. When the "A Jobs" batch gets near completion, many threads may not be processing any jobs until the "on_success" callback kicks off the next phase and creates the "B Jobs." In other words, we don't want almost idle CPUs as we transition from A to B to C.

Here's a rough diagram:

![image](https://user-images.githubusercontent.com/1118459/148191215-0f89bf8e-1418-4a9c-99ab-0a0e14bdaa04.png)

If I naively just created all the batches at once, I'd have say 10 threads creating the A jobs, 10 threads creating the B jobs, and 10 threads creating the C jobs. Thus, C jobs would be mixed in with A jobs, right, since they are all running concurrently? Instead, I want all the A jobs to run before the B jobs. I'm accomplishing that right now by creating the next workflow batch via the on_success callback.

I tried to create secondary batches and pass that to the jobs that do the bulk queuing, but I got an error as it's not feasible to create a batch that will later get some jobs.

**Can I create two parallel flows?**

**Comment by mperham** [comment]
> Instead, I want all the A jobs to run before the B jobs.

I'm confused. How can you have parallel workflows AND sequential execution?

You can create child batches within the overall batch which execute in parallel, e.g.

```ruby
overall = Sidekiq::Batch.new
overall.on(...)
overall.jobs do
  a = Sidekiq::Batch.new
  a.on(...)
  a.jobs do
    LoadJobs.perform_async(some_data)
  end
  b = Sidekiq::Batch.new
  b.on(...)
  b.jobs do
    LoadJobs.perform_async(some_other_data)
  end
end
```

The jobs loaded into A and B will run in parallel and get their own callbacks but the overall batch will get a callback when both child batches are done. That gives you the tools to make arbitrary workflows which are parallel or serial.

**Reply by justin808** [reply]
The problem with the above code is that the child jobs of batches A & B will be executed concurrently. I want the loading to go in order, and the running of the units of work to go in order.

**Comment by justin808** [comment]
> how can you have parallel workflows AND sequential execution?

We want to sequentially batch the jobs, using the batch loader. 

The "unit of work jobs" is the cache warming, such as server-side-rendering of a given URL. There will be about 120K of these.

The parallel workflows are:
1. creating the unit-of-work jobs via batch loading
2. doing the unit-of-work-jobs

Thus, per the image above, the green rectangles are the queries that create the jobs (that go into correspondingly-named batches). Each of those jobs will spawn many jobs to execute batch loading. When the batch loading of one type is done, I want the next type of batch loading to proceed. I figure the unit of work jobs will use a separate queue so as not to starve out the sequential jobs to coordinate batch loading.

Does that make sense?

**Reply by mperham** [reply]
If you want A jobs to strictly run before B jobs, you'll need to create the B jobs in A's callback. Once you create jobs, they are off and running.

**Reply by justin808** [reply]
Could we solve the problem by using a second queue with a lower priority for B jobs from B batches?

Suppose we have "batch loading" jobs within batches A1, A2, A3, A4. Could we have batch A1's callback trigger batch A2, and batch A2's callback trigger batch A3?

Batch A1 will be creating B1 batches with B1 jobs, and then batch A2 will be creating the B2 batch with B2 jobs. Since jobs from B2 batches are only added from A2 batches, we can be sure that A1 jobs get processed first. Correct?

**Reply by mperham** [reply]
A lower-priority queue is not a solution, it's a race condition. Batches need to **guarantee** the conditions for callbacks to run.

If A1 is a job, it doesn't have callbacks. If you are talking about child batches, that will work. Batch A1 can load jobs into itself and only kick off A2 once it is complete. Then you have Batch A kick off Batch B once all of its children are complete.

**Reply by justin808** [reply]
Hi @mperham, please see my updated write-up of the solution. I think this will work.

**Reply by mperham** [reply]
If A's callback creates B's jobs, there's no way to kick off B's jobs early while A's jobs are still processing. The closest thing you can get to that is to use the :complete callback so that if any of A's jobs fail, that failure won't block the kickoff of B's jobs.

**Reply by justin808** [reply]
The A jobs run very quickly, as they are creating the real work to be done. They are using the https://github.com/mperham/sidekiq/wiki/Batches#huge-batches feature. 

on_complete vs. on_success is a bit orthogonal, correct?

Or am I missing something?

What's a good example of an error where I'd still want to the flow to continue?

When I had (unknown) lost jobs due to scaling dynos and the on_success callbacks didn't run, the whole flow appeared stuck. I'm not sure if on_complete would run.

Is there any good discussion of error handling with batch callbacks? What happens with uncaught exceptions in the callbacks?

**Reply by mperham** [reply]
Batches don't have errors, jobs do. Jobs record their error and retry. A callback is just another job; you fix the error so the callback job runs successfully.

**Reply by justin808** [reply]
Sorry for the confusion: I updated the question:

> Is there any good discussion of error handling with batch callbacks? What happens with uncaught exceptions in the callbacks?

I meant what happens if there are errors raised within the callbacks?

Should all callback methods have a `rescue` at the end?

Or do you mean that the callback is run as a job, so it would work just like any other job that fails, and maybe then it shouldn't have a rescue?

And that all the answers are right here: https://github.com/mperham/sidekiq/wiki/Error-Handling#best-practices ?

**Reply by mperham** [reply]
As the Batches page says, every callback runs as a job. You do not need to rescue anything.
